proposals/002-text-utf-default.md

```diff
@@ -0,0 +1,156 @@
+    threshold.
+
+# People
+
+-   Performers:
+
+    -   Leader: Andrew Lelechenko (bodigrim)
+
+    -   Support: Emily Pillmore (emilypi)
+
+-   Reviewers:
+
+    -   The text maintainers
+
+        -   Xia Li-Yao (lysxia)
+
+        -   Emily Pillmore (emilypi)
+
+        -   Dan Cartwright (chessai)
+
+        -   Callan McGill (boarders)
+
+-   Stakeholders:
+
+    -   Edward Kmett: has been vocal about his use of `text-icu` and requires it not be broken.
+
+-   Ben Gamari: integration with GHC
+
+-   The Cabal maintainers (fgaz, emilypi, mikolaj): integration with Cabal
+
+Progress will be reported on a weekly basis to the HF Technical Agenda
+Track, with Emily as support for Andrew.
+
+# Timeline
+
+We expect that this project will take roughly 6 months to fully
+complete: 3-4 months to complete the code implementation, performance
+testing, and unit testing, another 1-2 months to integrate with
+stakeholders and diagnose any potential issues with the migration.
+
+**Preparation:**
+
+Using the HVR's existing [`text-utf8`](https://github.com/text-utf8) as
+a starting point, the following must be done before an implementation is
+started:
+
+-   Modernize the codebase and clear out the bitrot
+
+-   Establish a baseline for performance and any related issues.
+
+-   Update testing and performance benchmarks to make use of `inspection-testing` to ensure fusion is not broken in
+    subsequent UTF-8 related changes.
+
+An MVP should completely preserve standard user-facing API, and not
+break fusion. Performance should not significantly diverge from the
+existing UTF-16 text package. There will be an expected change to the
+exposed Text internals, in which case, breakage should be assessed by
+circulating a git commit reference to a release candidate as soon as
+possible. This candidate should be sourced publicly and loudly.
+
+**Implementation:**
+
+-   TBD: There is a straightforward implementation, but this one is left up to Andrew for comment.
+
+**Stakeholders:**
+
+-   Library authors will need to be made aware of changes and adjust accordingly. HF will provide a git reference to a complete MVP as
+    soon as possible, and produce a migration guide.
+
+-   In the case of GHC, Cabal, and other core infrastructure, we will work closely with these packages to help migrate and assess+fix
+    breakages.
+
+While we do not expect many authors to experience significant changes,
+there will be some help that needs to be given in terms of bumping
+Hackage bounds since this migration will be a major version bump. HF
+will need to coordinate with the Hackage Trustees to help move along
+packages that go out of date.
+
+# Deliverables
+
+-   text-2.0.0.0, which will provide a UTF-8 encoding for Text as a default for all versions going forward.
+
+-   A `text-utf16` package, which is a preservation of the current UTF-16 encoded text, for backwards compatibility.
```

Comment:
In reaching out to the community, is there demand for this support? If so, is there a reason to make this its own package, and not just e.g. a different datatype exported from the same `text` package?

Reply:
In theory the full matrix of
- UTF8, UTF16, UTF32 (LE, BE)
- sliced, non-sliced
- ByteArray#, pinned/unpinned, ForeignPtr backed

can be useful (c.f. avoiding copying etc.).

(for example, some people have been experimenting with parsing UTF32 backed text - it has random access, so it's great when inputs are of human size, re-encoding cost may be won in faster parsing).

Reply:
As long as your grammar do not have rules specific for non-ASCII characters (which is most often the case), one can parse UTF8-encoded data as a `ByteString`, benefiting from random access. Same as UTF32, but in 4x less memory. Note that this is not the case for UTF16, since it is not compatible with ASCII.

Reply:
If you parse from bytes, you should use `ByteString` (or something with name "bytes" in it). Then `Text` encoding is irrelevant.

Reply:
> If so, is there a reason to make this its own package, and not just e.g. a different datatype exported from the same `text` package?

Keeping both data types in `text` puts a double and infinitely-long burden on `text` maintainers. I don't think it's sustainable, and it's certainly most confusing from users' viewpoint. 

> In reaching out to the community, is there demand for this support? 

As covered in other sections, I suggest we leave this to the community: we are happy to bless and support anyone motivated enough to create a legacy utf16 fork and maintain it as long as desired. But we have neither sustainable resources nor strong motivation to take this burden exclusively ourselves.